keps/sig-node/2358-node-affinity-manager/README.md

```diff
@@ -0,0 +1,532 @@
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Node affinity field `RequiredDuringSchedulingRequiredDuringExecution` has been commented out since 2016 because it has not been implemented yet.
+Implementing it can provide a way to ensure pods are in the desired nodes at runtime.
+
+## Motivation
+
+In Kubernetes, "Node Affinity" controls how pods are scheduled. 
+However, there is currently no way to guarantee that the node affinity will still be met during runtime.
+
+### Goals
+
+- Uncomment the NodeAffinity `RequiredDuringSchedulingRequiredDuringExecution` field and corresponding validations.
+- Add a node affinity manager to the node lifecycle controllers to ensure pods being evicted if the selectors are no longer met at runtime.
+
+### Non-Goals
+
+## Proposal
+
+### User Stories
+
+#### Manage Nodes With Special Resources
+
+Assuming there are multiple nodes labeled with "app=cache", which means that cache apps like Redis can be scheduled on this node.
+I added some nodes to provide more available memory, and I wish to migrate all previous cache applications to these new nodes.
+With runtime NodeSelector constraints, I can simply delete the label from existing nodes to evict all cache apps.
+
+#### Manage Multi Tenant Cluster
+
+I have some nodes labeled with "userA=allow" and "userB=allow", which means that these nodes are available to user A and user B.
+As the scale grows, I added more nodes for user B, and all existing nodes are only available to user A.
+I can simply delete the "userB=allow" label from the existing nodes to re-schedule all pods of user B to these new nodes.
+
+#### Share Resources More Effectively
+
+Assuming I have some failure tolerance calculation apps that will continuously generate large temporary files, 
```

Comment:
Managing resource usage on the node doesn't sound like a good use case for node affinity. You can use something similar to what was done for taint based evictions to implement this use case: https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions